src/Behat/Behat/EventDispatcher/Cli/StopOnFailureController.php

```diff
@@ -82,28 +76,14 @@ public function execute(InputInterface $input, OutputInterface $output)
             return null;
         }
 
+        
         if ($input->getOption('strict')) {
-            $this->resultInterpretation = new StrictInterpretation();
+            $resultInterpretation = new StrictInterpretation();
+        } else {
+            $resultInterpretation = new SoftInterpretation();
         }
 
-        $this->eventDispatcher->addListener(ScenarioTested::AFTER, array($this, 'exitOnFailure'), -100);
-        $this->eventDispatcher->addListener(ExampleTested::AFTER, array($this, 'exitOnFailure'), -100);
+        $this->stopOnFailureHandler->registerListeners($resultInterpretation);
     }
 
-    /**
-     * Exits if scenario is a failure and if stopper is enabled.
-     *
-     * @param AfterScenarioTested $event
-     */
-    public function exitOnFailure(AfterScenarioTested $event)
```

Comment:
This looks like a BC break? Can we keep the method and proxy to the StopOnFailureHandler?

Reply:
Anything *can* be used outside of behat - user code is running quite tightly coupled with the behat code, and for example extensions have full access to reconfigure / use the DI container so we can't guarantee what people are doing.

I prefer to take a safe / strict view of semver unless things are explicitly marked as not part of the public API (for example, [phpunit is very clear about this](https://github.com/sebastianbergmann/phpunit/blob/921ca7c95d440437da5ce077cf691ebf90c116dc/src/Framework/Exception/Exception.php#L41))

So I would like to see the class constructor keep BC, as you say ideally with a comment to change it in future.

This `exitOnFailure` method is tricky though - it looks like it was only ever public because callbacks had to be in older PHP versions, and it was never intended to be called directly. So looking at it again I realise proxying to the new listener doesn't really make sense.

I think therefore it's extremely unlikely anyone will be calling that, and we could safely remove it.

Reply:
OK I will do that 👍

Reply:
I finally kept the event dispatcher in the constructor and add a setter as `@required` that is called via the DI.